server/src/main/java/org/opensearch/common/blobstore/BlobContainer.java

```diff
@@ -124,6 +127,33 @@ default long readBlobPreferredLength() {
      */
     void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException;
 
+    /**
+     * Used to check whether vendor plugin support for parallel upload of multiple streams is enabled or not.
+     * Returns false by default
+     *
+     * @return If multi-stream parallel uploads are supported
+     */
+    default boolean isMultiStreamUploadSupported() {
+        return false;
+    }
+
+    default boolean isRemoteDataIntegritySupported() {
+        return false;
+    }
+
+    /**
+     * Reads blob content from multiple streams, each from a specific part of the file, which is provided by the
+     * StreamContextSupplier in the WriteContext passed to this method. An {@link IOException} is thrown if reading
+     * any of the input streams fails, or writing to the target blob fails
+     *
+     * @param writeContext A WriteContext object encapsulating all information needed to perform the upload
+     * @return A {@link CompletableFuture} representing the upload
+     * @throws IOException if any of the input streams could not be read, or the target blob could not be written to
+     */
+    default CompletableFuture<Void> writeBlobByStreams(WriteContext writeContext) throws IOException {
```

Comment:
WriteContext seems too vague in the contract. Would prefer

```
default CompletableFuture<Void> writeBlobStreams(StreamContextSupplier streamContextSupplier, FileMetadata metadata, CheckedConsumer<Boolean, IOException> uploadFinalizer, WritePriority WritePriority) throws IOException {
```

Reply:
Splitting WriteContext would mean that in future if any other plugin implementer requires an additional attribute to perform the transfer then this would require signature changes which all other existing implementers have to adapt to.

Reply:
It's the same thing, adding an attribute in `WriteContext ` would require change on plugin implementor's side if done in one way(name remains unchanged however the contracts break and plugin won't compile). The idea is how better can you split so that now the contract is explicit and intuitive enough.

Reply:
We have a way of field type change without breaking the contract. We add a new field with a different name and desired type and deprecate the old one. This way we will also be able to provide enough time to plugins to migrate to the new field.

Reply:
We have seen in aws sdk migration in s3 repository plugin and ec2 plugin that how difficult and error prone it gets to change an externally shared contract. This is an attempt to handle those situations. I agree that this encapsulation is not revealing its properties intuitively. How about naming it something else than splitting it?

Reply:
If we build the contract right it should stand tall. Let's not worry about future use cases now. We can always have newer contracts and then deprecate the old ones. That's how most softwares upgrade themselves.
This shouldn't trade with a poor contract to start with

Reply:
I gave reference of v1-v2 migration because it started with replacing v1 sdk version property in OS with v2 version property (to maintain single SDK). Changes to one property (analogous to a param in your contract) required changes across multiple plugins. There could be potentially similar breaking changes for modifications made to a shared field as well depending on the role that field plays, right?

Let me add few references from AWS SDK as well where a single encapsulation is maintained. 
1. All public methods of [S3TransferManager](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/S3TransferManager.java#L162)
2. All abstractions of [S3AsyncClient](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/S3AsyncClient.html). There are few abstractions which have additional param(s) but they exist to provide overloaded version of their single request object based counterparts. Internally they invoke their single param counterparts only.
3. Similar abstractions are present in v1 SDK as well.

Request objects of these abstractions encapsulate file metadata, request body references (file or input stream or content providers), executors, listeners, etc.  

I am just providing a context of where I am coming from and why this was designed this way. I agree with you that we should design it in the right way and I am open to changes if you still strongly feel that we should split this.

Reply:
The examples you cite are those of client-server request-response models that interact over a pre-defined serialisation protocol which may not be the case here.

Reply:
Not exactly. These models are not transmitted over the wire and hence, they are not serializable and instead different properties of these objects such as the bucket name, object key, metadata, and content are serialized into an HTTP request by the client. 
They contain listeners, executors, flags, etc. which are used to manage the request lifecycle within the SDK itself.

Reply:
Don't want to get hung up on this. But if you look at the [S3AsyncClient](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/S3AsyncClient.html) example. All the requests are [SdkPojo](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/SdkPojo.html) with each [field](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/SdkField.html) with a mechanism to marshall/unmarshall. 
IMHO the interfaces in this context should be very clear and intuitive to define. I am fine doing a follow-up as well, but lets atleast agree on the right way forward

Reply:
Sorry for the confusion! I only checked S3 transfer manager abstractions. Those abstractions are composed of both non-serializable and serializable entities wrapped in one. For Primitive S3 SDKs (v1 and v2), as you said directly accept serializable objects. 
Tagging @raghuvanshraj to comment on the amount of work needed to split this. Based on this we can take up a follow up or plan to cover it as part of this PR itself. Does it seem fine?

Reply:
Splitting WriteContext can be taken as a follow up. I have captured it here: https://github.com/opensearch-project/OpenSearch/issues/8080